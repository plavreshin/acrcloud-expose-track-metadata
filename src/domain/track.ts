export type Track = {
  id: number;
  name: string;
  artistName: string;
  duration: number;
  isrc: string;
  releaseDate: string;
  createdAt: Date;
  updatedAt?: Date;
};

export type CreateTrack = {
  name: string;
  artistName: string;
  duration: number;
  isrc: string;
  releaseDate: string;
};

export type UpdateTrack = {
  name?: string;
  artistName?: string;
  duration?: number;
  isrc?: string;
  releaseDate?: string;
};
