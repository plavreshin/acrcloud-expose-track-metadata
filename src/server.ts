import * as dotenv from 'dotenv';
import * as http from 'http';
import GracefulShutdown from 'http-graceful-shutdown';

import { newApp } from './app';
import { initHttpGracefulShutdown } from './gracefulShutdown';

const envParsed = dotenv.config({ debug: true }).parsed!;

let server: http.Server;

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion

export const createServer = async (port: number = 9000) => {
  const configuration = {
    arcCloudToken: envParsed.ARC_CLOUD_TOKEN,
    secretKey: envParsed.SECRET_KEY,
  };
  newApp(configuration)
    .then(({ app }) => {
      // eslint-disable-next-line no-console
      console.log('Starting HTTP server on port %d', port);
      server = http.createServer(app).listen(port, () => {
        // eslint-disable-next-line no-console
        console.log('Listening on port %d', port);
      });

      server.on('close', () => {
        app.emit('server:stop');
        // eslint-disable-next-line no-console
        console.log('HTTP server close.');
      });

      const shutdownOptions: GracefulShutdown.Options = {
        signals: 'SIGINT SIGTERM',
        timeout: 30000,
        development: process.env.NODE_ENV === 'development',
        finally: () => {
          // eslint-disable-next-line no-console
          console.log('Server gracefully shutted down.....');
        },
      };
      initHttpGracefulShutdown(server, app, shutdownOptions);
    })
    .catch((err) => {
      // eslint-disable-next-line no-console
      console.log('Error loading app: %s', err);
      process.exit(1);
    });
};

createServer();
