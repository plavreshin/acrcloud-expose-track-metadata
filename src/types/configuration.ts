export type AppConfiguration = {
  arcCloudToken: string;
  secretKey: string;
};
