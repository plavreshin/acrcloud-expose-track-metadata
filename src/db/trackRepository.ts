import * as O from 'fp-ts/lib/Option';
import { Option } from 'fp-ts/lib/Option';
import { Database } from 'sqlite';
import { CreateTrack, Track } from 'src/domain/track';

export type TrackModel = {
  id: number;
  name: string;
  artistName: string;
  duration: number;
  isrc: string;
  releaseDate: string;
  createdAt: Date;
  updatedAt?: Date;
};

export class TrackRepository {
  private readonly db: Database;

  constructor(db: Database) {
    this.db = db;
  }

  public async getTracks(): Promise<Track[]> {
    return this.db.all<TrackModel[]>('SELECT * FROM track');
  }

  public async getTrackByNameAndArtist(name: string, artistName: string): Promise<Track[]> {
    return this.db.all<TrackModel[]>(
      'SELECT * FROM track WHERE name like ? and artistName like ?',
      `%${name}%`,
      `%${artistName}%`,
    );
  }

  public async persistTrack(track: CreateTrack): Promise<Option<number>> {
    const now = new Date();
    const result = await this.db.run(
      `INSERT INTO track (name, artistName, duration, isrc, releaseDate, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [
        track.name,
        track.artistName,
        track.duration,
        track.isrc,
        track.releaseDate,
        now.toISOString(),
        now.toISOString(),
      ],
    );
    return O.fromNullable(result.lastID);
  }

  public async updateTrack(id: number, track: Track): Promise<Option<Track>> {
    const now = new Date();
    const result = await this.db.run(
      `UPDATE track SET name = ?, artistName = ?, duration = ?, isrc = ?, releaseDate = ?, updatedAt = ? WHERE id = ?`,
      [track.name, track.artistName, track.duration, track.isrc, track.releaseDate, now.toISOString(), id],
    );
    return result.changes && result.changes > 0 ? O.some(track) : O.none;
  }

  public async getTrackById(id: number): Promise<Option<Track>> {
    const row = await this.db.get<TrackModel>('SELECT * FROM track WHERE id = ?', id);
    return O.fromNullable(row);
  }

  public async getTrackByIds(ids: number[]): Promise<Track[]> {
    return this.db.all<TrackModel[]>(`SELECT * FROM track WHERE id IN (${ids.map((_) => '?').join(',')})`, ids);
  }

  public async deleteTrack(id: number): Promise<boolean> {
    const result = await this.db.run('DELETE FROM track WHERE id = ?', id);
    return result.changes ? result.changes > 0 : false;
  }
}
