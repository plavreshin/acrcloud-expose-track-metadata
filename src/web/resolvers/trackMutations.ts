import * as E from 'fp-ts/lib/Either';
import { pipe } from 'fp-ts/lib/function';
import { GraphQLError } from 'graphql';
import { log } from 'src/common/logger';
import { UpdateTrack } from 'src/domain/track';

import { Resolvers } from '../__generated__/resolvers-types';

export const trackMutationResolvers: Resolvers = {
  Mutation: {
    updateTrack: async (_, args, context) => {
      log.info(`updateTrack: ${args.id} by ${context.auth.email}`);
      const { id, name, artistName, duration, isrc, releaseDate } = args;
      const updateTrack: UpdateTrack = {
        name: name ?? undefined,
        artistName: artistName ?? undefined,
        duration: duration ?? undefined,
        isrc: isrc ?? undefined,
        releaseDate: releaseDate ?? undefined,
      };
      const trackE = await context.services.trackService.updateTrackMetadata(id, updateTrack);
      return pipe(
        trackE,
        E.fold(
          (err) => {
            log.warn(`No track found for ${args.id}`, err);
            throw new GraphQLError(err, {
              extensions: { code: err },
            });
          },
          (track) => track,
        ),
      );
    },
    deleteTrack: async (_, args, context) => {
      log.info(`deleteTrack: ${args.id} by ${context.auth.email}`);
      return context.services.trackService.deleteTrack(args.id);
    },
  },
};
