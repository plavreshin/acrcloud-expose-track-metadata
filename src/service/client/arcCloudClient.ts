// import fetch from '@adobe/node-fetch-retry';
import { none, Option, some } from 'fp-ts/lib/Option';
import { log } from 'src/common/logger';
import { AppConfiguration } from 'src/types/configuration';

export type ArtisEntry = {
  name: string;
};

export type Album = {
  name: string;
  release_date: string;
};

export type TrackMetadata = {
  name: string;
  artists: ArtisEntry[];
  isrc: string;
  duration_ms: number;
  album: Album;
};

type ArcCloudTrackLookupRequest = {
  track: string;
  artists: string[];
};

type TrackMetadataResponse = {
  data: TrackMetadata[];
};

export class ArcCloudClient {
  private readonly config: AppConfiguration;

  private metadataEndpoint = 'https://eu-api-v2.acrcloud.com/api/external-metadata/tracks';

  constructor(config: AppConfiguration) {
    this.config = config;
  }

  public async getTrackMetadata(name: string, artistName: string): Promise<Option<TrackMetadata[]>> {
    try {
      const response = await this.getTrackMetadataFromArcCloud(name, artistName);

      log.info(`Arc cloud returned status ${response.status} for ${name}`);
      if (response.status === 200) {
        const json = await response.json();
        const trackMetadataResponse = json as TrackMetadataResponse;
        return trackMetadataResponse?.data?.length > 0 ? some(trackMetadataResponse.data) : none;
      }

      log.error(`Arc cloud returned status ${response.status} for ${name}`);
      return none;
    } catch (err) {
      log.error('Failed to get track metadata from 3rd party service for name' + name, err);
      return Promise.reject(err);
    }
  }

  protected async getTrackMetadataFromArcCloud(name: string, artistName: string) {
    const payload: ArcCloudTrackLookupRequest = {
      track: name,
      artists: [artistName],
    };

    const queryParams = new URLSearchParams();
    queryParams.append('format', 'json');
    queryParams.append('query', JSON.stringify(payload));

    const url = new URL(this.metadataEndpoint);
    url.search = queryParams.toString();

    return fetch(url, {
      headers: {
        Authorization: `Bearer ${this.config.arcCloudToken}`,
      },
      method: 'GET',
    });
  }
}
