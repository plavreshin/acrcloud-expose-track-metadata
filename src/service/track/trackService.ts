import * as A from 'fp-ts/lib/Array';
import * as E from 'fp-ts/lib/Either';
import { Either } from 'fp-ts/lib/Either';
import { pipe } from 'fp-ts/lib/function';
import { Option } from 'fp-ts/lib/Option';
import * as O from 'fp-ts/lib/Option';
import { log } from 'src/common/logger';
import { TrackRepository } from 'src/db/trackRepository';
import { CreateTrack, Track, UpdateTrack } from 'src/domain/track';
import { ValidationError } from 'src/types/validation';

import { ArcCloudClient, TrackMetadata } from '../client/arcCloudClient';

export class TrackService {
  private readonly trackRepository: TrackRepository;
  private readonly arcCloudClient: ArcCloudClient;

  constructor(trackRepository: TrackRepository, arcCloudClient: ArcCloudClient) {
    this.trackRepository = trackRepository;
    this.arcCloudClient = arcCloudClient;
  }

  public async getTracks(): Promise<Track[]> {
    return this.trackRepository.getTracks();
  }

  public async getTrackById(id: number): Promise<Option<Track>> {
    return this.trackRepository.getTrackById(id);
  }

  public async getTrackMetadataOrRetrieve(
    name: string,
    artistName: string,
  ): Promise<E.Either<ValidationError, Track[]>> {
    const knownTracks = await this.trackRepository.getTrackByNameAndArtist(name, artistName);

    if (knownTracks.length === 0) {
      log.info(`Track with name ${name} not found in DB, retrieving from 3rd party service`);
      const tracksMetadata: Either<ValidationError, TrackMetadata[]> = await this.arcCloudClient
        .getTrackMetadata(name, artistName)
        .then((tracks) => {
          return E.fromOption(() => ValidationError.NotFound)(tracks);
        })
        .catch((err) => {
          log.error(`No tracks match ${name} due to afailure in ArcCloud request`, err);
          return E.left(ValidationError.ExternalApiError);
        });

      const createTracksE = pipe(
        tracksMetadata,
        E.map((tracksMetadata) => {
          log.info(`Found ${JSON.stringify(tracksMetadata)} tracks for name ${name}`);
          const createTracks = tracksMetadata.reduce((acc, item) => {
            for (const artist of item.artists) {
              if (artist.name === artistName) {
                acc.push({
                  name: item.name,
                  artistName: artist.name,
                  duration: item.duration_ms,
                  isrc: item.isrc,
                  releaseDate: String(item.album.release_date),
                });
              }
            }
            return acc;
          }, [] as CreateTrack[]);
          return createTracks;
        }),
      );

      return pipe(
        createTracksE,
        E.fold(
          async (err) => {
            return Promise.resolve(E.left(err));
          },
          async (createTracks) => {
            log.info(`Found ${createTracks.length} tracks for name ${name}, persisting to DB`);
            const promises = createTracks.map((track) => this.trackRepository.persistTrack(track));
            const idsOpt = await Promise.all(promises);
            const ids = pipe(idsOpt, A.compact);
            const tracks = await this.trackRepository.getTrackByIds(ids);
            return Promise.resolve(E.right(tracks));
          },
        ),
      );
    }

    return Promise.resolve(E.right(knownTracks));
  }

  public async updateTrackMetadata(id: number, update: UpdateTrack): Promise<E.Either<ValidationError, Track>> {
    const track = await this.trackRepository.getTrackById(id);

    return pipe(
      track,
      O.fold(
        async () => {
          return Promise.resolve(E.left(ValidationError.NotFound));
        },
        async (track) => {
          const updatedTrack: Track = {
            ...track,
            name: update.name ?? track.name,
            artistName: update.artistName ?? track.artistName,
            isrc: update.isrc ?? track.isrc,
            duration: update.duration ?? track.duration,
            releaseDate: update.releaseDate ?? track.releaseDate,
            updatedAt: new Date(),
          };

          const updatedO = await this.trackRepository.updateTrack(id, updatedTrack);
          return E.fromOption(() => ValidationError.NotFound)(updatedO);
        },
      ),
    );
  }

  public async deleteTrack(id: number): Promise<boolean> {
    return this.trackRepository.deleteTrack(id);
  }
}
