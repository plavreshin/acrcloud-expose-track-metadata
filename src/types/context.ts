import { TrackService } from 'src/service/track/trackService';

export type AuthPayload = {
  email: string;
};

export interface Context {
  auth: {
    email: string;
  };
  services: {
    trackService: TrackService;
  };
}
