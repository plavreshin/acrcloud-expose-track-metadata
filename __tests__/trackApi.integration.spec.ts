import { ApolloServer } from '@apollo/server';
import { readFileSync } from 'fs';
import path from 'path';
import { Database } from 'sqlite';
import { resolvers as scalarResolvers, typeDefs as scalarTypeDefs } from 'graphql-scalars';
import { AppConfiguration } from 'src/types/configuration';
import { TrackRepository } from 'src/db/trackRepository';
import { ArcCloudClient } from 'src/service/client/arcCloudClient';
import { TrackService } from 'src/service/track/trackService';
import { Context } from 'src/types/context';
import { trackQueryResolvers } from 'src/web/resolvers/trackQueries';
import { Track } from 'src/domain/track';
import { trackMutationResolvers } from 'src/web/resolvers/trackMutations';
import { none, some } from 'fp-ts/lib/Option';
import { right } from 'fp-ts/lib/Either';

const getAllTracksQuery = `
query TrackQuery {
  getAllTracks {
    id
    name
    artistName
  }
}
`;

const getSingleTrackQuery = `
query TrackQuery {
  getTrackById(id: 1) {
      id
      name
      artistName
    }  
  }
`;

const getTrackByNameAndArtistQuery = `
query TrackQuery {
  getTrackByNameAndArtist(name: "Summer", artistName: "Calvin Harris") {
    id
    createdAt
    updatedAt
    name
    artistName
    duration
    isrc
    releaseDate
  }
}
`;

const deleteTrackMutation = `
mutation TrackMutation {
  deleteTrack(id: 1)
  }
`;

const updateTrackMutation = `
mutation UpdateTrack {
  updateTrack(id: 1, name: "Updated") {
    name
  }
}
`;

const tracks: Track[] = [
  {
    id: 1,
    name: 'name',
    artistName: 'artistName',
    duration: 1,
    isrc: 'isrc',
    releaseDate: '2024-01-01',
    createdAt: new Date(),
  },
];

describe('trackApi', () => {
  const config: AppConfiguration = {
    arcCloudToken: 'token',
    secretKey: 'track-api',
  };

  const trackRepositoryMock: TrackRepository = new TrackRepository({} as Database);
  const arcCloudClientMock: ArcCloudClient = new ArcCloudClient(config);
  const trackService: TrackService = new TrackService(trackRepositoryMock, arcCloudClientMock);

  const contextValue = {
    auth: {
      email: 'email@email.me',
    },
    services: {
      trackService,
    },
  };

  const makeServer = () => {
    const typeDefs = readFileSync(path.join(path.resolve(), './src/web/schema.graphql'), 'utf-8');

    const server = new ApolloServer<Context>({
      typeDefs: [...scalarTypeDefs, typeDefs],
      resolvers: [scalarResolvers, trackQueryResolvers, trackMutationResolvers],
    });

    return server;
  };

  it('should return existing all tracks successfully', async () => {
    trackService.getTracks = jest.fn().mockResolvedValue(Promise.resolve(tracks));

    const res = await makeServer().executeOperation(
      {
        query: getAllTracksQuery,
      },
      {
        contextValue: {
          auth: {
            email: 'email@email.me',
          },
          services: {
            trackService,
          },
        },
      },
    );

    expect(res).toMatchSnapshot();
  });

  it('should return single track successfully', async () => {
    trackService.getTrackById = jest.fn().mockResolvedValue(some(tracks[0]));

    const res = await makeServer().executeOperation(
      {
        query: getSingleTrackQuery,
        variables: {
          id: 1,
        },
      },
      {
        contextValue,
      },
    );

    expect(res).toMatchSnapshot();
  });

  it('should return track by name and artist successfully', async () => {
    trackService.getTrackMetadataOrRetrieve = jest.fn().mockResolvedValue(right(tracks));

    const res = await makeServer().executeOperation(
      {
        query: getTrackByNameAndArtistQuery,
        variables: {
          name: 'Summer',
          artistName: 'Calvin Harris',
        },
      },
      {
        contextValue,
      },
    );

    expect(res).toMatchSnapshot();
  });

  it('should delete track', async () => {
    trackService.deleteTrack = jest.fn().mockResolvedValue(true);

    const res = await makeServer().executeOperation(
      {
        query: deleteTrackMutation,
      },
      {
        contextValue: contextValue,
      },
    );

    expect(res).toMatchSnapshot();
  });

  it('should get validation error when track is not found', async () => {
    trackService.getTrackById = jest.fn().mockResolvedValue(none);

    const res = await makeServer().executeOperation(
      {
        query: getSingleTrackQuery,
      },
      {
        contextValue: contextValue,
      },
    );

    expect(res).toMatchSnapshot();
  });

  it('should update track name', async () => {
    trackService.updateTrackMetadata = jest.fn().mockResolvedValue(right(tracks[0]));

    const res = await makeServer().executeOperation(
      {
        query: updateTrackMutation,
      },
      {
        contextValue: contextValue,
      },
    );

    expect(res).toMatchSnapshot();
  });
});
