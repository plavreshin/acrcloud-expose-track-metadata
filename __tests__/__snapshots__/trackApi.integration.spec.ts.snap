// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trackApi should delete track 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "deleteTrack": true,
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`trackApi should get validation error when track is not found 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "getTrackById": null,
      },
      "errors": [
        {
          "extensions": {
            "code": "NotFound",
          },
          "locations": [
            {
              "column": 3,
              "line": 3,
            },
          ],
          "message": "No track found for 1",
          "path": [
            "getTrackById",
          ],
        },
      ],
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`trackApi should return existing all tracks successfully 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "getAllTracks": [
          {
            "artistName": "artistName",
            "id": 1,
            "name": "name",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`trackApi should return single track successfully 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "getTrackById": {
          "artistName": "artistName",
          "id": 1,
          "name": "name",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`trackApi should return track by name and artist successfully 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "getTrackByNameAndArtist": [
          {
            "artistName": "artistName",
            "createdAt": 2024-01-04T16:37:35.598Z,
            "duration": 1,
            "id": 1,
            "isrc": "isrc",
            "name": "name",
            "releaseDate": "2024-01-01",
            "updatedAt": null,
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`trackApi should update track name 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "updateTrack": {
          "name": "name",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
